### PR TITLE
[py12] core tests

### PR DESCRIPTION
**Pull request by can-anyscale**
Add core unit-tests for python 3.12 version - all are passing now w00t w00t, thanks to all the great work from @hongchaodeng 

Test:
- CI

### CODE REVIEW

File: python/ray/tests/accelerators/test_intel_gpu.py [hunk 1]
```diff
@@ -103,6 +103,10 @@ def test_set_current_process_visible_accelerator_ids():
 
 
 if __name__ == "__main__":
+    # Intetionally skip this test on Python 3.12
```
Comment (jjyao) [inline python/ray/tests/accelerators/test_intel_gpu.py:106]: Could you add a comment why we skip?

File: python/ray/tests/test_dashboard_profiler.py [hunk 2]
```diff
@@ -249,4 +249,8 @@ def get_actor_stack():
 
 
 if __name__ == "__main__":
+    # TODO(can): No py-spy for python 3.12
+    if sys.version_info >= (3, 12):
```
Comment (jjyao) [inline python/ray/tests/test_dashboard_profiler.py:253]: We normally use pytest.skip to skip tests

**Comment by can-anyscale** [comment]
address @jjyao's comments
- use @@pytest.mark.skipif, except for the `test_tensorflow` file because I need to skip the whole file for python 3.12 (we haven't upgraded tensorflow for python 3.12 yet). pytest natively supports this via `pytest.skip(allow_module_level=true)` but that doesn't play too well with our pytest+bazel setup
- added reasons to skip test_npu and test_intel_gpu as they are from external contributors, @hongchaodeng if there are issues we can add to the reason let me know

**Review by hongchaodeng — APPROVED** [review]
LGTM

**Comment by can-anyscale** [comment]
let's go!

**Pull request merged**